### PR TITLE
feat: resolve import specifiers through a project's declared SvelteKit aliases

### PR DESCRIPTION
**Pull request by oekazuma**
## Why

`resolveRepoLocalPath` understood exactly two specifier forms: `$lib/…`, hard-coded to `src/lib/…`, and relative paths. Everything else returned `undefined`, which every rule built on it reads as "not a project-local module — stay silent".

That was recorded as a light pre-1.0 constraint. Measurement retired the estimate: in a real monorepo of several SvelteKit apps, a custom alias is used more widely than `$lib`, and `$lib` is forbidden there by convention. For such a project the affected rules do not under-report a little — `security/shared-state-import` is **inert**, because every import it examines has to resolve to a project-local path before the rule can see it at all.

There is a second, worse form of the same gap. `$lib` is not a constant: `kit.files.lib` moves it. A project that moves it got `$lib/…` resolved to `src/lib/…` anyway — not a miss but a **wrong answer**, pointing at a file that is not the one imported.

## What

Aliases are read from `svelte.config.{js,ts}` (`kit.alias`, plus `kit.files.lib` for `$lib`) during fact collection, compiled into an ordered list, and stored as `Project.kitAliases`. `resolveRepoLocalPath` takes that list and resolves against it.

The one design decision everything else follows from:

> **The resolver reproduces what the bundler does, mechanism included — not a cleaner scheme that usually agrees with it.**

SvelteKit's `get_config_aliases` builds an ordered array with `$lib` prepended, and the resolver over it takes the **first** match, not the best one. Reproducing that mechanism rather than approximating it is what makes three otherwise-plausible designs wrong:

| Plausible design | Why it is wrong |
| --- | --- |
| Longest key wins (deterministic, order-independent) | `{ '$a': 'src/x', '$a/b': 'src/y' }` resolves `$a/b/c` to `src/x/b/c` at build time — the `$a/b` entry is unreachable. Longest-key answers `src/y/c`: a different, possibly existing file, fed to default-on security rules. |
| A user's `kit.alias.$lib` overrides `kit.files.lib` | Kit prepends its own `$lib` entry, so `files.lib` always wins and a user `$lib` entry is dead. |
| Drop an alias whose value is not a string literal | Under first-match-wins a shorter list does not answer less, it answers **differently** — a later entry answers in the dropped entry's place, resolving a specifier that is `undefined` today. Such an entry now holds its position and its mode, and answers `undefined` when reached. |

Kit compiles three match modes, not one, and the third is the one a from-scratch design misses: declaring both `'x'` and `'x/*'` narrows `'x'` to an exact match, so the two entries partition the space instead of the first swallowing everything. Modes are assigned from the **declared** key set, never from the subset whose values happened to be readable.

Deliberately unresolved rather than guessed at: a computed alias value (`packages/core` runs no user code and performs no I/O, so it cannot be evaluated); an alias object containing a spread or a computed **key**, where an unknown key at a known position could shadow anything after it with no `find` to record that with — the whole map is discarded, keeping only the `$lib` entry that nothing can shadow; a literal absolute value; and a project whose SvelteKit options are passed to the `sveltekit()` Vite plugin, since Kit ignores `svelte.config` entirely in that case.

Collection adds **no I/O**: both channels already read the config for `kit.paths.base` and threw the source away. `packages/cli/test/io-budget.test.ts` passes with its numbers untouched, which is the mechanical proof.

## User-visible effect

Two of the affected rules are on by default, so a project that changes nothing can see new findings after this lands — findings the rules always owed and could not see. `security/shared-state-import` is the one whose behaviour visibly changes; `security/handler-state-write` gains only its `.set()`/`.update()` shape. `architecture/private-scope-import` is opt-in and gains reach only where declared. `performance/heavy-import` is unaffected: it matches raw specifiers against bare package names and never resolves a project-local path.

## Notes

- The lib `server/` exemption in `isLocalStateSpecifier` now derives its directory from the resolved `$lib` rather than hard-coding `src/lib/server`. Without that, a project moving `files.lib` would get a **new critical false positive** — the exemption that exists to allow DB/KV singletons would silently stop applying to them.
- `resolveComponentPath` (`packages/cli/src/providers/source/resolve.ts`) is a second specifier mapping, still hard-coded to `$lib/`, driving head resolution. Extending alias resolution to it is out of scope here; the doc comment that claimed alias support "must stay a single-site change" has been corrected to say so rather than left as a false claim.

## Verification

`core` 1096, `cli` 779, `vite` 205, `mcp` 25 tests pass; typecheck clean in all four packages; lint and format clean. The end-to-end test drives a declared alias from a config source through real fact collection into an actual finding, with a negative control differing only in whether the alias is declared — without it, every other test would pass over a resolver that silently never matched.

Design: `docs/superpowers/specs/2026-07-30-kit-alias-resolution-design.md`. This unblocks `architecture/route-component-import`, whose own design is already approved and sequenced next.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Import resolution now supports project-defined SvelteKit aliases from `svelte.config.{js,ts}`, including customized `$lib` locations.
  * Analysis can now trace aliased imports for improved privacy/security detection, using SvelteKit precedence (first-match-wins).
  * Added alias-aware module import scanning across CLI and Vite flows.

* **Documentation**
  * Expanded English and Japanese guidance on supported aliases and explicitly documented unresolved-alias scenarios.

* **Tests**
  * Added/updated CLI and Vite integration tests, plus core regression and edge-case coverage for unresolved and opaque alias values.

### CODE REVIEW

File: docs/src/content/docs/guides/(setup)/configuration.mdx [hunk 1]
```diff
@@ -188,6 +188,25 @@ section with its exact option names and defaults.
   (directory glob → allowed child names), `unitScopes` (root glob → the names a unit's children may
   take) and `exclude`. Off until one of the two is set.
 
+### Import aliases
+
+Rules that follow imports (`architecture/private-scope-import`, `security/shared-state-import`,
+`security/handler-state-write`) resolve specifiers through the aliases your project declares in
+`svelte.config.js` — `kit.alias`, plus `kit.files.lib` when `$lib` has been moved. They are read
+statically, in the same order SvelteKit builds them, and the first matching alias wins, exactly as it
+does at build time.
```
[the remark this answers is not in the document]
Comment (oekazuma) [reply docs/src/content/docs/guides/(setup)/configuration.mdx:197]: Fixed in 19111106. `SVELTE_CONFIG_FILES` reads both, so naming only the `.js` form understated it. Corrected in the English guide, the Japanese guide, and the changeset.

File: docs/src/content/docs/guides/(setup)/configuration.mdx [hunk 2]
```diff
@@ -188,6 +188,25 @@ section with its exact option names and defaults.
...
+Three cases are not resolved:
+
+- an alias whose value is computed (`path.resolve(...)`, a template literal) rather than a plain string —
+  the specifier is treated as one we cannot see, so nothing is reported for it;
+- a project that passes its SvelteKit options to the `sveltekit()` plugin in `vite.config.ts` instead of
+  `svelte.config.js` — same result, nothing reported;
+- a `kit.alias` object containing a spread (`{ ...shared, '$a': 'src/a' }`) or a computed key
+  (`{ [key]: 'src/a' }`) — an unknown key could shadow any entry declared after it, so the _whole_ of
+  `kit.alias` is discarded rather than just that one entry. Every literal alias declared alongside it stops
+  resolving too (`$lib` is unaffected, since nothing can shadow the position it always occupies).
```
[the remark this answers is not in the document]
Comment (oekazuma) [reply docs/src/content/docs/guides/(setup)/configuration.mdx]: Fixed in 19111106, and the list had drifted further than this comment says — it was missing two cases, not one:

- a **literal absolute target** (`'/opt/shared/src'`, and the drive-letter form raised in the `kit-module-parse.ts` comment);
- an **unreadable `kit.files.lib`**, which is a new case as of the same commit — see the reply on the `filesLib` comment.

I dropped the leading "Three cases" count rather than changing it to "Five". `AGENTS.md` warns that hard-coded counts rot on every addition, and this list has now grown twice in one PR. Both languages updated.

File: docs/superpowers/plans/2026-07-30-kit-alias-resolution.md [hunk 3]
```diff
@@ -0,0 +1,1147 @@
...
+  /**
+   * `kit.alias` entries in declaration order, `value: null` where the config's value is not a
+   * string literal. **Undefined means the key set is unknowable** — a spread or a computed key
+   * puts an unknown key at a known position, and an unknown key could shadow anything after it,
+   * with no `find` to record that with. The caller then discards every user entry.
+   */
+  entries?: { key: string; value: string | null }[];
+  /** `kit.files.lib`, when it is a string literal. */
+  filesLib?: string;
+};
```
[the remark this answers is not in the document]
Comment (oekazuma) [reply docs/superpowers/plans/2026-07-30-kit-alias-resolution.md:513]: Good catch — fixed in 19111106. This was an asymmetry in my own design, not just a missing case: the PR already gives an unreadable `kit.alias` **value** an opaque entry (`replacement: null`, which matches and holds its position but cannot resolve), and I never extended that treatment to `$lib` itself. `filesLibOf` collapsed "no `lib` property" and "a `lib` property we cannot read" into the same `undefined`, and `raw.filesLib ?? 'src/lib'` then silently answered with the default — a wrong answer for a project that moved its lib directory behind an expression.

`RawKitAliases.filesLib` is now `string | null | undefined` with the three states documented, and `compileKitAliases` compiles the unreadable state to `{ find: '$lib', replacement: null, match: 'prefix' }`.

**I took the opposite decision on the exemption, though.** The suggestion is to "not apply the `src/lib/server` exemption when `$lib` is opaque" — read literally that means not exempting, i.e. **reporting**, which is a critical false positive in a default-on rule. With the lib root unknown we cannot tell whether a resolved path is one of the legitimate singletons the exemption exists for, so `isLocalStateSpecifier` returns `false` and the rule stays silent. That follows the design doc's stated asymmetry — an unresolved specifier is conservative, a misresolved one is a false positive — and matches what the function already does for any specifier it cannot resolve. The cost is a missed finding in a rare configuration, which is the side of the trade the spec chooses.

One note on the test: `$lib/…` specifiers already resolve to `undefined` once `$lib` is opaque, so a test using them would pass without the exemption guard existing. The covering test uses a **relative** specifier that lands inside the unknown lib `server/` directory, which is the only path where the guard is load-bearing.

The design doc now records both decisions.

File: docs/superpowers/specs/2026-07-30-route-component-import-design.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+**Charter row:** verdict #1 — importing a route component (L1)
+**Blocked on:** SvelteKit alias resolution, which measurement moved ahead of this rule. See
+"Why this waits" below.
+
+## The problem
+
+A SvelteKit route entry — `+page.svelte`, `+layout.svelte`, `+error.svelte` — is written on the
+assumption that Kit renders it. Kit hands a page its `data` and `params`; it hands an error page its
+`page.error` and `page.status`. Imported from somewhere else, the component receives none of that and
+renders against nothing.
```
[the remark this answers is not in the document]
Comment (oekazuma) [reply docs/superpowers/specs/2026-07-30-route-component-import-design.md:14]: Checked this against the installed SvelteKit (2.70.1) rather than the docs, and the sentence is correct as written — `params` **is** a route-component prop.

Kit's own generated types say so:

```js
// @sveltejs/kit/src/core/sync/write_types/index.js:279
'export type PageProps = { params: RouteParams; data: PageData; form: ActionData }'
```

and the generated root passes it to every page and layout:

```svelte
<!-- @sveltejs/kit/src/core/sync/write_root.js:43 -->
<Pyramid bind:this={components[depth]} data={d} {form} params={page.params}>
```

(also lines 49, 61, 62, 72, 75, 86, 88 — every branch of that template).

`page.params` from `$app/state` is the *other* way to reach the same value, and the one to use deeper in a component tree, but it is not the only one. Props were added for `params` in SvelteKit 2.16, which is likely why the search result says otherwise.

No change made. The spec sentence stays: an imported route component receives none of what Kit would have passed it, and `params` is part of that.

File: docs/superpowers/specs/2026-07-30-route-component-import-design.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+### The check, in four steps
+
+1. **Skip type-only imports.** A `import type P from './+page.svelte'` is erased at build; nothing
+   renders, so the harm the rule describes cannot occur. Reporting it would be a false positive.
+2. **Resolve the specifier to a repo-relative path.** `$lib/` and relative specifiers resolve; a bare
+   package does not.
+3. **Require the resolved path to be a route entry**: under the routes directory, with a basename
+   matching `/^\+(page|layout)(@[^./]*)?\.svelte$/` or equal to `+error.svelte`. Seven names, the
+   `@` forms included, because a breakout entry is an entry.
```
[the remark this answers is not in the document]
Comment (oekazuma) [reply docs/superpowers/specs/2026-07-30-route-component-import-design.md]: Correct, and this one would have shipped as a rule that silently skips valid route entries — fixed in 19111106.

Confirmed against the installed SvelteKit (2.70.1). `analyze()` strips **only** the component extension before matching:

```js
// @sveltejs/kit/src/core/sync/create_manifest_data/index.js:530-531
const name = file.slice(0, -component_extension.length);
const pattern = /^\+(?:(page(?:@(.*))?)|(layout(?:@(.*))?)|(error))$/;
```

so `+page@foo.bar.svelte` gives `name === '+page@foo.bar'` and `uses_layout === 'foo.bar'` — a layout name containing dots is a legal breakout target, since it is just a route directory name. The narrower `@[a-zA-Z0-9_-]*` in the same file applies to the non-component `+page.server.js` family, where a dot would be ambiguous with the `.server` suffix; it does not constrain `.svelte` components.

The spec now uses `(@.*)?`. That also falsified a claim two paragraphs below it: the spec said "**Seven names**, the `@` forms included", which only held while the suffix was a closed set — with `@(.*)` the set is unbounded, so it is now stated as the shapes it actually is. A dotted-`@`-suffix case was added to the spec's Testing section.

Documentation-only: the rule this specifies is not implemented yet, which is why the correction lands in the spec rather than in a matcher.

File: packages/core/src/kit-module-parse.ts [hunk 6]
```diff
@@ -466,76 +467,142 @@ function normalizePosix(path: string): string | undefined {
...
+    // plugin resolves with entries.find(), so position is precedence. An entry that matches
+    // but carries no readable value stops here rather than letting a later entry answer.
+    const entry = aliases.find((a) => aliasMatches(a, spec));
+    if (entry?.replacement == null) return undefined;
+    // An absolute replacement (a literal `/opt/shared/src`, not `path.resolve(...)` — that
+    // form is non-literal and already opaque) is outside the project by definition. Left
+    // unchecked, `normalizePosix` drops the leading empty segment from `/opt/...` and
+    // answers `opt/shared/src/...` — a project-relative path that names a different file.
+    if (entry.replacement.startsWith('/')) return undefined;
+    path = entry.replacement + spec.slice(entry.find.length);
```
[the remark this answers is not in the document]
Comment (oekazuma) [reply packages/core/src/kit-module-parse.ts:535]: Fixed in 19111106, with the suggested regex. The reachability is worth spelling out: values are posixified before this guard runs, so `'C:\shared\src'` arrives as `C:/shared/src` — the normalisation step this PR added is what put a Windows absolute path in front of a check that only knew about `/`. Test added alongside the existing absolute-target case.
